### DISCUSSION TITLE
Starlette path convertor ignored in favor of the function's signature

### DISCUSSION THREAD

**Discussion by AzariasB** [fastapi/fastapi#8621] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [x] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import APIRouter

router = APIRouter(prefix="/users")

@router.get(
    "",
    name=f"get_all_users",
)
def get_user():
    return []

@router.get(
    "/{user_id:int}",
    name=f"get_user",
)
def get_user(user_id):
    print(user_id)
    return user_id
```

### Description

I want to use to starlette path parameter convertor because it allows me to dynamically define the parameter type (whereas, as far as I know, I can't dynamically declare a python parameter type)

There are several problems when using the Starlette path parameters convertors that are documented [here](https://www.starlette.io/routing/).

The swagger documentation does not indicate the type of the parameter
![image](https://user-images.githubusercontent.com/10064282/157638230-9ab2838e-d8f9-4ad2-8040-0d6975a2244f.png)

Here's what I would expect when declaring the convertor in the url :
![image](https://user-images.githubusercontent.com/10064282/157639364-e3747052-7a6c-4569-805a-a02aaa2f9d06.png)

When querying with incorrect parameters (passing in a string instead of an int for exemple : `/users/bob` instead of `/users/1`), the server does not throw an error because of the wrongly typed parameter, it just passes the parameter to the function as a string.

It seems like FastAPI only relies on the type of the function, and ignores the convertors declared in the url. If there was a way to force certain routes to always rely on the url informations instead of the function's parameters type, that would be great !

### Operating System

Linux

### Operating System Details

N/A

### FastAPI Version

0.75.0

### Python Version

3.9.7

**Comment by YuriiMotov** [comment, marked as answer]
Yes, convertors are used by Starlette to match path. And, FastAPI relies on type annotations to generate schema and validate parameters.


And, regarding:

> When querying with incorrect parameters (passing in a string instead of an int for exemple : /users/bob instead of /users/1), the server does not throw an error because of the wrongly typed parameter, it just passes the parameter to the function as a string.

This doesn't work this way in current version. If you specify `"/{user_id:int}"` and try passing `/users/bob` you will get `404` error response
